File: BMWCenter/Core/Care/CareCoordinator.swift
import Foundation
import SwiftData
import Combine
import UIKit
import CoreLocation

@MainActor
final class CareCoordinator: ObservableObject {
    @Published private(set) var activeChips: [ActiveAlert] = []
    @Published private(set) var readiness: Double = 0
    @Published private(set) var isEngineReady = false
    @Published private(set) var readinessLabel: String?
    @Published private(set) var liveEcoScore: Double = 100
    @Published private(set) var oilTempC: Double?
    @Published private(set) var oilIsEstimated = true
    @Published private(set) var thermalCountdownS: Double?
    @Published private(set) var tripCardImage: UIImage?
    @Published private(set) var showTripCard = false

    let baseline: BaselineLearner
    let scheduler: CueScheduler
    let engineReady: EngineReadyService
    let coldShield: ColdEngineShield
    let overheat: OverheatWatchdog
    let thermal: ThermalShockGuard
    let battery: BatteryGuardian
    let eco: EcoCoach
    let gear: GearCoach
    let challenges: ChallengeEngine
    let streak: StreakService
    let badges: BadgeService
    let trim: FuelTrimMonitor
    let thermostat: ThermostatWatch
    let maintenance: AdaptiveMaintenance
    let cardRenderer = TripSummaryCardRenderer()

    private let settings: AppSettings
    private let modelContext: ModelContext
    private var features: [CareFeature] = []
    private var listenTask: Task<Void, Never>?
    private var context = CareContext()
    private var cleanWarmups = 0
    private var tripCount = 0

    init(settings: AppSettings, modelContext: ModelContext, announcer: AudioAnnouncer) {
        self.settings = settings
        self.modelContext = modelContext
        self.baseline = BaselineLearner(modelContext: modelContext)
        self.scheduler = CueScheduler(announcer: announcer)
        self.engineReady = EngineReadyService()
        self.coldShield = ColdEngineShield(modelContext: modelContext)
        self.overheat = OverheatWatchdog(baseline: baseline, modelContext: modelContext)
        self.thermal = ThermalShockGuard(modelContext: modelContext)
        self.battery = BatteryGuardian(modelContext: modelContext)
        self.eco = EcoCoach()
        self.gear = GearCoach()
        self.challenges = ChallengeEngine(modelContext: modelContext, coldShield: coldShield)
        self.streak = StreakService(modelContext: modelContext)
        self.badges = BadgeService(modelContext: modelContext)
        self.trim = FuelTrimMonitor(baseline: baseline, modelContext: modelContext)
        self.thermostat = ThermostatWatch(baseline: baseline, modelContext: modelContext)
        self.maintenance = AdaptiveMaintenance(baseline: baseline, modelContext: modelContext)

        features = [
            coldShield, engineReady, overheat, thermal, battery,
            eco, gear, challenges, trim, thermostat, maintenance
        ]

        scheduler.onPresented = { [weak self] cue, plan in
            guard let self else { return }
            if plan.phoneChip {
                let alert = ActiveAlert(
                    id: cue.id,
                    title: cue.chipTitle ?? cue.text,
                    severity: SeverityRouter.alertSeverity(from: cue.severity)
                )
                if !self.activeChips.contains(where: { $0.id == alert.id }) {
                    self.activeChips.insert(alert, at: 0)
                    if self.activeChips.count > 6 {
                        self.activeChips = Array(self.activeChips.prefix(6))
                    }
                }
            }
        }
        applyCueFrequency()
    }

    func startListening(to snapshots: AsyncStream<VehicleSnapshot>, supportedPIDs: @escaping () -> Set<UInt8>) {
        listenTask?.cancel()
        listenTask = Task { [weak self] in
            var lastEval = Date.distantPast
            for await snap in snapshots {
                guard let self else { break }
                let now = Date()
                guard now.timeIntervalSince(lastEval) >= 1.0 else { continue }
                lastEval = now
                await MainActor.run {
                    self.evaluate(snapshot: snap, supportedPIDs: supportedPIDs(), now: now)
                }
            }
        }
    }

    func stop() {
        listenTask?.cancel()
        listenTask = nil
    }

    func evaluate(snapshot: VehicleSnapshot, supportedPIDs: Set<UInt8>, now: Date = .now) {
        applyCueFrequency()
        maintenance.airflowEnabled = settings.careAirflowWatch

        // Oil temperature: measured OBD PID only — no estimation in UI or Care context.
        oilTempC = snapshot.oilTempC
        oilIsEstimated = snapshot.oilTempC == nil

        context.now = now
        context.supportedPIDs = supportedPIDs
        context.ambientC = snapshot.ambientC ?? context.ambientC
        context.oilTempC = snapshot.oilTempC
        context.oilIsEstimated = snapshot.oilTempC == nil
        context.fuelType = settings.fuelType
        context.isTurbo = settings.isTurbo
        context.sensitivityOffsetC = sensitivityOffset()
        context.isVehicleStopped = (snapshot.speedKmh ?? 0) < 2
        context.tripDistanceKm = 0
        context.lastRefuelAt = nil

        var allCues: [CareCue] = []
        for feature in features {
            guard feature.isEnabled(settings: settings) else { continue }
            if !feature.isAvailable(supportedPIDs: supportedPIDs), !supportedPIDs.isEmpty {
                continue
            }
            allCues.append(contentsOf: feature.evaluate(snapshot: snapshot, context: &context))
        }

        readiness = engineReady.readiness
        isEngineReady = engineReady.isReady
        readinessLabel = engineReady.remainingLabel
        liveEcoScore = eco.liveScore
        thermalCountdownS = thermal.countdownSeconds

        if settings.careSpokenCues {
            scheduler.enqueueAll(allCues, now: now)
        } else {
            // Still surface critical/protective chips
            for cue in allCues where cue.severity >= .protective {
                scheduler.enqueue(cue, now: now)
            }
        }
    }

    func onTripEnded(_ trip: Trip) {
        context.tripDistanceKm = trip.distanceKm
        context.tripDurationS = trip.durationS
        context.isVehicleStopped = true

        var endCues: [CareCue] = []
        for feature in features {
            guard feature.isEnabled(settings: settings) else { continue }
            endCues.append(contentsOf: feature.onTripEnded(trip: trip, context: context))
        }
        scheduler.enqueueAll(endCues, now: .now)

        let hadViolation = coldShield.coldViolationsThisTrip > 0
        if !hadViolation { cleanWarmups += 1 }
        tripCount += 1
        streak.recordDay(
            score: trip.scoreTotal ?? 0,
            hadProtectionViolation: hadViolation
        )
        _ = badges.evaluateAwards(
            isStopped: true,
            cleanWarmups: cleanWarmups,
            compliantCooldowns: 0,
            harshAccelKm: 0,
            tripCount: tripCount,
            longHaulKm: trip.distanceKm,
            thermostatCaught: thermostat.caughtFault
        )

        let cleanWarmup = !hadViolation
        if settings.careTripSummaryCard, cardRenderer.shouldRender(distanceKm: trip.distanceKm) {
            Task { [weak self] in
                try? await Task.sleep(nanoseconds: 3_000_000_000)
                await self?.presentTripCard(trip, cleanWarmup: cleanWarmup)
            }
        }

        coldShield.resetTrip()
        engineReady.resetTrip()
        eco.resetTrip()
    }

    func resetBaselines() {
        baseline.reset()
    }

    private func presentTripCard(_ trip: Trip, cleanWarmup: Bool) {
        let harsh = (trip.events ?? []).filter { $0.typeRaw == "harshBrake" }.count
        var start: CLLocationCoordinate2D?
        if let lat = trip.startLatitude, let lon = trip.startLongitude {
            start = CLLocationCoordinate2D(latitude: lat, longitude: lon)
        }
        var end: CLLocationCoordinate2D?
        if let lat = trip.endLatitude, let lon = trip.endLongitude {
            end = CLLocationCoordinate2D(latitude: lat, longitude: lon)
        }
        let model = TripSummaryCardRenderer.CardModel(
            vehicleName: settings.vehicleName,
            date: trip.endedAt ?? trip.startedAt,
            distanceKm: trip.distanceKm,
            durationS: trip.durationS,
            avgL100: trip.avgL100,
            cost: trip.fuelUsedL * settings.pricePerLiter,
            currencyCode: settings.currencyCode,
            score: trip.scoreTotal,
            bestStretchKm: nil,
            bestStretchL100: nil,
            cleanWarmup: cleanWarmup,
            properCooldown: true,
            harshBrakes: harsh,
            startCoord: start,
            endCoord: end,
            hideLocation: settings.careHideLocationSharing
        )
        tripCardImage = cardRenderer.renderImage(model: model)
        showTripCard = tripCardImage != nil
    }

    private func sensitivityOffset() -> Double {
        switch settings.careSensitivity {
        case "early": return -2
        case "calm": return 2
        default: return 0
        }
    }

    private func applyCueFrequency() {
        switch settings.careCueFrequency {
        case "low": scheduler.updateFrequency(1.5)
        case "high": scheduler.updateFrequency(0.7)
        default: scheduler.updateFrequency(1.0)
        }
    }
}

File: mobile/src/core/care/careCoordinator.ts
import { create } from "zustand";
import { overheatWatchdog } from "./overheatWatchdog";
import { coldEngineShield } from "./coldEngineShield";
import { thermalShockGuard } from "./thermalShockGuard";
import { batteryGuardian } from "./batteryGuardian";
import { fuelTrimMonitor } from "./fuelTrimMonitor";
import { thermostatWatch } from "./thermostatWatch";
import { engineReadyService } from "./engineReadyService";
import { ecoCoach } from "./ecoCoach";
import { gearCoach } from "./gearCoach";
import { adaptiveMaintenance } from "./adaptiveMaintenance";
import { challengeEngine } from "./challengeEngine";
import { streakService } from "./streakService";
import { badgeService } from "./badgeService";
import { cueScheduler } from "./cueScheduler";
import { emptyCareContext } from "./careTypes";
import { alertSeverityFrom } from "./severityRouter";
import { buildTripSummaryCard, shouldRenderTripCard, type TripSummaryCardModel } from "./tripSummaryCard";
import { useAppSettings } from "../settings/appSettings";
import { effectivePricePerLiter } from "../fuel/effectivePrice";
import type { VehicleSnapshot } from "../obd/vehicleSnapshot";
import type { Trip } from "../storage/models";
import type { CareCue, CareContext } from "./careTypes";
import type { ActiveAlert } from "../alerts/alertTypes";
import type { CareFeature } from "./careFeature";

const FEATURES: CareFeature[] = [
  coldEngineShield,
  engineReadyService,
  overheatWatchdog,
  thermalShockGuard,
  batteryGuardian,
  ecoCoach,
  gearCoach,
  challengeEngine,
  fuelTrimMonitor,
  thermostatWatch,
  adaptiveMaintenance,
];

interface CareCoordinatorState {
  activeChips: ActiveAlert[];
  readiness: number;
  isEngineReady: boolean;
  readinessLabel?: string;
  liveEcoScore: number;
  thermalCountdownS?: number;
  fullScreenCue?: CareCue;
  tripSummaryCard?: TripSummaryCardModel;

  evaluate: (snapshot: VehicleSnapshot, now?: number) => void;
  onTripEnded: (trip: Trip) => Promise<void>;
  dismissFullScreenCue: () => void;
  dismissTripCard: () => void;
}

let context: CareContext = emptyCareContext();
let cleanWarmups = 0;
let tripCount = 0;
let lastEval = 0;

/** Coolant/overheat protection cues are safety-critical and stay on even with "Enable
 * alerts" off — mirrors the forced full-screen/spoken overheat override elsewhere. */
function isForcedOverheatCue(cue: CareCue): boolean {
  return cue.id.startsWith("overheat.") && (cue.severity === "protective" || cue.severity === "critical");
}

/** "Enable alerts" is the master switch: off means no chips/cues/notifications from the
 * Care coordinator either, not just the built-in AlertEngine rules — except the forced
 * overheat safety cue above. */
function filterForEnableAlerts(cues: CareCue[], settings: { enableAlerts: boolean }): CareCue[] {
  return settings.enableAlerts ? cues : cues.filter(isForcedOverheatCue);
}

cueScheduler.onPresented = (cue, plan) => {
  const isForcedOverheat = cue.id.startsWith("overheat.") && (cue.severity === "protective" || cue.severity === "critical");
  if (plan.fullScreen || isForcedOverheat) {
    useCareCoordinator.setState({ fullScreenCue: cue });
  }
  if (plan.phoneChip) {
    const alert: ActiveAlert = { id: cue.id, title: cue.text, severity: alertSeverityFrom(cue.severity) };
    useCareCoordinator.setState((s) => {
      if (s.activeChips.some((a) => a.id === alert.id)) return s;
      return { activeChips: [alert, ...s.activeChips].slice(0, 6) };
    });
  }
};

function applyCueFrequency() {
  const settings = useAppSettings.getState();
  switch (settings.careCueFrequency) {
    case "low":
      cueScheduler.updateFrequency(1.5);
      break;
    case "high":
      cueScheduler.updateFrequency(0.7);
      break;
    default:
      cueScheduler.updateFrequency(1.0);
  }
}

function sensitivityOffset(): number {
  const settings = useAppSettings.getState();
  if (settings.careSensitivity === "early") return -2;
  if (settings.careSensitivity === "calm") return 2;
  return 0;
}

export const useCareCoordinator = create<CareCoordinatorState>((set, get) => ({
  activeChips: [],
  readiness: 0,
  isEngineReady: false,
  liveEcoScore: 100,

  evaluate: (snapshot, now = Date.now()) => {
    if (now - lastEval < 1000) return;
    lastEval = now;

    applyCueFrequency();
    const settings = useAppSettings.getState();

    context = {
      ...context,
      now,
      ambientC: snapshot.ambientC ?? context.ambientC,
      oilTempC: snapshot.oilTempC,
      oilIsEstimated: snapshot.oilTempC == null,
      sensitivityOffsetC: sensitivityOffset(),
      isVehicleStopped: (snapshot.speedKmh ?? 0) < 2,
    };

    const allCues: CareCue[] = filterForEnableAlerts(
      FEATURES.flatMap((feature) => (feature.isEnabled(settings) ? feature.evaluate(snapshot, context) : [])),
      settings
    );

    if (settings.careSpokenCues) {
      cueScheduler.enqueueAll(allCues, now);
    } else {
      allCues.filter((c) => c.severity === "protective" || c.severity === "critical").forEach((c) => cueScheduler.enqueue(c, now));
    }

    set({
      readiness: engineReadyService.readiness,
      isEngineReady: engineReadyService.isReady,
      readinessLabel: engineReadyService.remainingLabel,
      liveEcoScore: ecoCoach.liveScore,
      thermalCountdownS: thermalShockGuard.countdownSeconds,
    });
  },

  onTripEnded: async (trip) => {
    const settings = useAppSettings.getState();
    context = { ...context, tripDistanceKm: trip.distanceKm, tripDurationS: trip.durationS, isVehicleStopped: true };

    const endCues: CareCue[] = [];
    for (const feature of FEATURES) {
      if (!feature.isEnabled(settings) || !feature.onTripEnded) continue;
      endCues.push(...(await feature.onTripEnded(trip, context)));
    }
    cueScheduler.enqueueAll(filterForEnableAlerts(endCues, settings), Date.now());

    const hadViolation = coldEngineShield.coldViolationsThisTrip > 0;
    if (!hadViolation) cleanWarmups += 1;
    tripCount += 1;

    await streakService.recordDay(trip.scoreTotal ?? 0, hadViolation);
    await badgeService.evaluateAwards({
      isStopped: true,
      cleanWarmups,
      compliantCooldowns: 0,
      harshAccelKm: 0,
      tripCount,
      longHaulKm: trip.distanceKm,
      thermostatCaught: false,
    });

    if (settings.careTripSummaryCard && shouldRenderTripCard(trip.distanceKm)) {
      const model = buildTripSummaryCard(trip, settings.vehicleName, effectivePricePerLiter(), settings.currencyCode, !hadViolation, settings.careHideLocationSharing);
      setTimeout(() => useCareCoordinator.setState({ tripSummaryCard: model }), 1500);
    }

    coldEngineShield.resetTrip();
    engineReadyService.resetTrip();
    ecoCoach.resetTrip();
    thermalShockGuard.resetTrip();
  },

  dismissFullScreenCue: () => set({ fullScreenCue: undefined }),
  dismissTripCard: () => set({ tripSummaryCard: undefined }),
}));
